"Building a Playlist Generator with Sentence Transformers"

NimaBoscarino opened the pull request:
This is a small post to wrap up the Playlist Generator project with a brief technical rundown of how ST and Gradio were used. I've put it in the HF communications calendar for Wednesday July 13th.

I still need to make a diagram (marked as TO DO in the blog content) on Monday.

I've also adjusted the post number + the associated paths for the previous post (85 -> 86).

osanseviero commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+
+A short while ago I published a [playlist generator](https://huggingface.co/spaces/NimaBoscarino/playlist-generator) that I’d built using **[Sentence Transformers](http://sbert.net)** and [Gradio](https://gradio.app), and I followed that up with a [reflection on how I try to use my projects as effective learning experiences](https://huggingface.co/blog/your-first-ml-project). But how did I actual ...
+
+As we’ve explored in [previous posts on the Hugging Face blog](https://huggingface.co/blog/getting-started-with-embeddings), Sentence Transformers (ST) is a library that gives us tools to generate sentence embeddings, which have a variety of uses. Since I had access to a dataset of song lyrics, I decided to leverage ST’s **[semantic search](https://www.sbert.net/examples/applications/semantic-sea ...
```
I would avoid having too many links, as that might lead to people going to other sites and not coming back. From here I think you could remove semantic search link Same below, I would remove the introduction to blocks are you link it in two places in the post

osanseviero commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+
+[The ST documentation highlights many of the choices](https://www.sbert.net/docs/pretrained_models.html), along with their evaluation metrics and some descriptions of their intended use-cases. The **[MS MARCO models](https://www.sbert.net/docs/pretrained-models/msmarco-v5.html)** are trained on Bing search engine queries, but since they also perform well on other domains I decided any one of thes ...
+
+Each model in ST has a configurable [input sequence length](https://www.sbert.net/examples/applications/computing-embeddings/README.html#input-sequence-length) (up to a maximum), after which your inputs will be truncated. The model I chose had a max sequence length of 512 word pieces, which, as I found out, is often not enough to embed entire songs. Luckily, there’s an easy way for us to split ly ...
```
I would remove the doc link here

osanseviero commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+
+** TODO: Diagram with Figma **
+
+To actually generate the embeddings, we can call the the `.encode()`  method on our Sentence Transformers model and pass it our list of strings. Then we can save our embeddings however we like – in this case I opted to pickle them. To be able to share our embeddings with others, we can even **[upload the Pickle file to a Hugging Face dataset](https://huggingface.co/blog/getting-started-with-embed ...
```
```suggestion
To actually generate the embeddings, you can call the the `.encode()` method of the Sentence Transformers model and pass it a list of strings. Then you can save the embeddings however you like – in this case I opted to pickle them. To be able to share you embeddings with others, you can even **[upload the Pickle file to a Hugging Face dataset](https://huggingface.co/blog/getting-started-with-embeddings#2-host-embeddings-for-free-on-the-hugging-face-hub)**.
```

osanseviero commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+verses = [...] # Load up your strings in a list
+corpus_embeddings = embedder.encode(verses, show_progress_bar=True)
+
+with open('verse-embeddings.pkl', "wb") as fOut:
+        pickle.dump(corpus_embeddings, fOut)
```
Here there is a double indent

osanseviero replied:
Maybe it's worth mentioning how to upload this to the Hub as you load them from the Hub in a later section

osanseviero commented on playlist-generator-st.md:
````diff
@@ -0,0 +1,152 @@
+# You can use the "corpus_id" to look up the original song
+```
+
+Since we’re searching for any verse that matches the text prompt, there’s a good chance that the semantic search will find multiple verses from the same song. One simple way to avoid duplicates is by increasing the number of verse embeddings that `util.semantic_search` fetches for us with the `top_k` parameter, and then we can drop duplicate songs as we map each verse embedding to its respective  ...
````
```suggestion
Since we’re searching for any verse that matches the text prompt, there’s a good chance that the semantic search will find multiple verses from the same song. One simple way to avoid duplicates is by increasing the number of verse embeddings that `util.semantic_search` fetches with the `top_k` parameter, and then you can drop duplicate songs as you map each verse embedding to its respective song. Experimentally, I found that when I set `top_k=20` I rarely get duplicates!
```

abidlabs replied:
I didn't really understand this paragraph. Why does increasing the number of search results in fewer duplicates? It should result in more duplicates? Do you mean to say that when you set `top_k=20`, you tend to get at least 9 distinct songs after remove duplicates?

NimaBoscarino replied:
Yeah good call, the way I phrased this is pretty confusing. Yup, that's what I mean – I'll adjust it!

osanseviero commented on playlist-generator-st.md:
````diff
@@ -0,0 +1,152 @@
+lyrics = pd.read_csv(hf_hub_download("NimaBoscarino/playlist-generator-private", repo_type="dataset", filename="lyrics_new.csv", use_auth_token=auth_token))
+```
+
+The Gradio Blocks API lets you build *multi-step* interfaces, which means that you’re free to create quite complex sequences for your demos. We’ll take a look at some example code snippets here, but [check out the project code to see it all in action](https://huggingface.co/spaces/NimaBoscarino/playlist-generator/blob/main/app.py). For this project, we want users to choose a text prompt and then, ...
````
```suggestion
The Gradio Blocks API lets you build *multi-step* interfaces, which means that you’re free to create quite complex sequences for your demos. We’ll take a look at some example code snippets here, but [check out the project code to see it all in action](https://huggingface.co/spaces/NimaBoscarino/playlist-generator/blob/main/app.py). For this project, we want users to choose a text prompt and then, after the semantic search is complete, users should have the ability to choose a song from the results to inspect the lyrics. With Gradio, this can be built iteratively by starting off with defining the initial input components and then registering a `click` event on the button. There’s also a `Radio` component, which will get updated to show the names of the songs for the playlist.
```

osanseviero commented on playlist-generator-st.md:
````diff
@@ -0,0 +1,152 @@
+)
+```
+
+This way, when the button gets clicked, Gradio grabs the current value of the TextArea and passes it to a function, shown below:
````
```suggestion
This way, when the button gets clicked, Gradio grabs the current value of the `TextArea` and passes it to a function, shown below:
```

osanseviero commented on playlist-generator-st.md:
````diff
@@ -0,0 +1,152 @@
+    )
+```
+
+In that function, we use the text prompt to conduct the semantic search. As we see above, to push updates to the Gradio components in the app, our function just needs to return components created with the `.update()` method. Since we connected the `song_option` Radio component to `fetch_songs.click` with its `output` parameter, `generate_playlist` can control the choices for our Radio component!
````
```suggestion
In that function, we use the text prompt to conduct the semantic search. As seen above, to push updates to the Gradio components in the app, the function just needs to return components created with the `.update()` method. Since we connected the `song_option` `Radio` component to `fetch_songs.click` with its `output` parameter, `generate_playlist` can control the choices for the `Radio `component!
```

osanseviero commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+
+In that function, we use the text prompt to conduct the semantic search. As we see above, to push updates to the Gradio components in the app, our function just needs to return components created with the `.update()` method. Since we connected the `song_option` Radio component to `fetch_songs.click` with its `output` parameter, `generate_playlist` can control the choices for our Radio component!
+
+We can even do something similar to the Radio component, in order to let users choose which song lyrics to view. [Visit the code on Hugging Face Spaces to see it in detail!](https://huggingface.co/spaces/NimaBoscarino/playlist-generator/blob/main/app.py)
```
```suggestion
You can even do something similar to the `Radio` component in order to let users choose which song lyrics to view. [Visit the code on Hugging Face Spaces to see it in detail!](https://huggingface.co/spaces/NimaBoscarino/playlist-generator/blob/main/app.py)
```

osanseviero reviewed (commented):
Very cool blog post :fire: Thank you!

abidlabs commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+    </a>
+</div>
+
+A short while ago I published a [playlist generator](https://huggingface.co/spaces/NimaBoscarino/playlist-generator) that I’d built using **[Sentence Transformers](http://sbert.net)** and [Gradio](https://gradio.app), and I followed that up with a [reflection on how I try to use my projects as effective learning experiences](https://huggingface.co/blog/your-first-ml-project). But how did I actual ...
```
Yes! A quick guide on embedding: https://gradio.app/embedding_gradio_demos/

NimaBoscarino replied:
Good idea! I'll embed it with a media query to hide it on small screens, since otherwise the layout breaks. (Just like I did here: https://github.com/huggingface/blog/blob/9b9aaa117437ba290ecac1dd6a8ea86b2011a249/your-first-ml-project.md?plain=1#L74)

abidlabs commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+
+As we’ve explored in [previous posts on the Hugging Face blog](https://huggingface.co/blog/getting-started-with-embeddings), Sentence Transformers (ST) is a library that gives us tools to generate sentence embeddings, which have a variety of uses. Since I had access to a dataset of song lyrics, I decided to leverage ST’s **[semantic search](https://www.sbert.net/examples/applications/semantic-sea ...
+
+We’ll be looking at a slightly advanced use of Gradio, so if you’re a beginner to the library I recommend exploring the [Gradio guides](https://gradio.app/guides/) and the [Introduction to Blocks](https://gradio.app/introduction_to_blocks/) before tackling the Gradio-specific parts of this post. Also, note that while I won’t be releasing the lyrics dataset, the **[lyrics embeddings are available  ...
```
```suggestion
We’ll be looking at a slightly advanced use of Gradio, so if you’re a beginner to the library I recommend reading the [Introduction to Blocks](https://gradio.app/introduction_to_blocks/) before tackling the Gradio-specific parts of this post. Also, note that while I won’t be releasing the lyrics dataset, the **[lyrics embeddings are available on the Hugging Face Hub](https://huggingface.co/datasets/NimaBoscarino/playlist-generator)** for you to play around with. Let’s jump in! 🪂
```

abidlabs commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+
+fetch_songs = gr.Button(value="Generate Your Playlist!")
+
+song_option = gr.Radio(interactive=True)
```
You shouldn't need `interactive=True` here
```suggestion
song_option = gr.Radio()
```

abidlabs commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,152 @@
+In that function, we use the text prompt to conduct the semantic search. As we see above, to push updates to the Gradio components in the app, our function just needs to return components created with the `.update()` method. Since we connected the `song_option` Radio component to `fetch_songs.click` with its `output` parameter, `generate_playlist` can control the choices for our Radio component!
+
+We can even do something similar to the Radio component, in order to let users choose which song lyrics to view. [Visit the code on Hugging Face Spaces to see it in detail!](https://huggingface.co/spaces/NimaBoscarino/playlist-generator/blob/main/app.py)
+
```
Is there a reason you don't mention the grid of Images that you create? FYI for the grid of images, you could use the `Gallery` component, which makes outputting 9 images a LOT easier: https://gradio.app/docs/#gallery

NimaBoscarino replied:
Yup! The Gallery component is awesome, but I ran into this bug while I was building the project: https://github.com/gradio-app/gradio/issues/1590

NimaBoscarino replied:
Oh and I avoided going into the grid of images here because I was trying to focus on just the stuff for hooking up multi-step apps. I'd written it originally with the image stuff and it got too bloated

abidlabs commented:
Awesome blog post @NimaBoscarino 🔥

omarespejel reviewed (approved):
LGTM after the changes proposed by Omar S.! :)

NimaBoscarino commented:
Thanks so much for reviewing!! I've addressed the comments, and I've also added a figure to illustrate the embedding process. Let me know if there are any more note! Otherwise, I'll merge this on Wednesday morning (Pacific Time) 😊

osanseviero commented on playlist-generator-st.md:
```diff
@@ -0,0 +1,172 @@
+---
```
I would rename this file to remove the `-st` part. Shorter links are better :fire:

osanseviero reviewed (approved):
Looks great! Let's ship it :fire:
